src/combinatorics/simple_graph/hom.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2022 Joanna Choules. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joanna Choules
+-/
+import topology.category.Top.limits
+import combinatorics.simple_graph.subgraph
+
+/-!
+# Homomorphisms from finite subgraphs
+
+This file defines the type of finite subgraphs of a `simple_graph` and proves a compactness result
+for homomorphisms to a finite codomain.
+
+## Main statements
+
+* `simple_graph.exists_hom_of_all_finite_homs`: If every finite subgraph of a (possibly infinite)
+  graph `G` has a homomorphism to some finite graph `F`, then there is also a homomorphism `G →g F`.
+
+## Notations
+
+`→fg` is a variant on `→g` where the domain is a finite subgraph of some supergraph `G`.
+
+## Implementation notes
+
+The proof here uses compactness as formulated in `nonempty_sections_of_fintype_inverse_system`. For
+finite subgraphs `G'' ≤ G'`, the inverse system `fin_subgraph_hom_functor` restricts homomorphisms
+`G' →fg F` to domain `G''`.
+-/
+
+universes u v
+variables {V : Type u} {W : Type v} {G : simple_graph V} {F : simple_graph W}
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+  { G' : G.subgraph // G'.verts.finite }
+
+/-- A graph homomorphism from a finite subgraph of G to F. -/
+abbreviation fin_subgraph_hom (G' : G.fin_subgraph) (F : simple_graph W) := G'.val.coe →g F
+infix ` →fg ` : 50 := fin_subgraph_hom
+
+/-- The finite subgraph of G generated by a single vertex. -/
+def vertex_singleton (v : V) : G.fin_subgraph :=
+ ⟨{verts := {v},
+   adj := λ _ _, false,
+   adj_sub := λ _ _, by contradiction,
+   edge_vert := λ _ _, by contradiction},
+  by simp only [set.finite_singleton]⟩
+
+/-- The finite subgraph of G generated by a single edge. -/
+def edge_singleton {u v : V} (e : G.adj u v) : G.fin_subgraph :=
+  ⟨{verts := {u, v},
+    adj := λ x y, (x = u ∧ y = v) ∨ (y = u ∧ x = v),
+    adj_sub := λ x y, by {rintro (⟨rfl, rfl⟩ | ⟨rfl, rfl⟩), exact e, exact e.symm},
+    edge_vert := λ x _, by {rintro (⟨rfl, _⟩ | ⟨_, rfl⟩); simp}},
+  by simp only [set.finite.insert, set.finite_singleton]⟩
```

Comment:
I have a branch that defines these two graphs (not as `fin_subgraph`s though). I'll create a PR soon, merge in some of this PR, and add you as a co-author.

Naming-wise, the temporary names I came up with were `simple_graph.singleton_subgraph` and `simple_graph.subgraph_of_adj`, but your names seem good too.